src/os/bluestore/BlueStore.cc

```diff
@@ -1744,6 +1744,64 @@ bool BlueStore::Blob::put_ref(
   return b.release_extents(empty, logical, r);
 }
 
+bool BlueStore::Blob::try_reuse_blob(uint32_t min_alloc_size,
+                		     uint32_t target_blob_size,
+		                     uint32_t b_offset,
+		                     uint32_t *length0) {
+  assert(min_alloc_size);
+  assert(target_blob_size);
+  if (!get_blob().is_mutable()) {
+    return false;
+  }
+
+  uint32_t length = *length0;
+  uint32_t end = b_offset + length;
+
+  // Currently for the sake of simplicity we omit blob reuse if data is
+  // unaligned with csum chunk. Later we can perform padding if needed.
+  if (get_blob().has_csum() &&
+     ((b_offset % get_blob().get_csum_chunk_size()) != 0 ||
+      (end % get_blob().get_csum_chunk_size()) != 0)) {
+    return false;
+  }
+
+  auto blen = get_blob().get_logical_length();
+  uint32_t new_blen = blen;
+  if (b_offset >= blen) {
+    //new data totally stands out of the existing blob
+    new_blen = b_offset + length;
+  } else {
+    //new data overlaps with the existing blob
+    new_blen = MAX(blen, length + b_offset);
+    if (!get_blob().is_unallocated(
+      b_offset,
+      new_blen > blen ? blen - b_offset : length)) {
+          return false;
+    }
+  }
+  if (new_blen > blen) {
```

Comment:
If i'm understanding this correctly, the new_blen > blen case can only happen if we take the top branch above (b_offset >= blen).  If we take the second path, that means we're writing into an unallocated AU inside a larger blob.  And our write always falls within one AU (this is _do_write_small), so we can't possibly be extending the blob.  Right?